piet-cairo/src/lib.rs

```diff
@@ -245,20 +245,30 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
                 match format {
                     ImageFormat::Rgb => {
                         for x in 0..width {
-                            data[dst_off + x * 4 + 0] = buf[src_off + x * 3 + 2];
-                            data[dst_off + x * 4 + 1] = buf[src_off + x * 3 + 1];
-                            data[dst_off + x * 4 + 2] = buf[src_off + x * 3 + 0];
+                            write_rgb(
```

Comment:
Might it be simpler for us to use `#[cfg(target_endian = "big")]` here and have two distinct codepaths? I'm not totally sure this makes sense, it really comes down to what the generated code looks like, but curious if that's an approach you had considered.

Reply:
No, I hadn't really considered that since letting `u32::to_ne_bytes()` deal with endianess seemed like the simplest thing to me. Also, using `#[cfg]` means that this is extra code which might need actual test runs after updating. With the `to_ne_bytes()`-approach, I'd feel more confident at changing the code since most of the code is not endianess-specific.

Anyway, do I understand correctly that you'd prefer something like the following? (No testing done at all - I did not even try building this on little endian)
```diff
diff --git a/piet-cairo/src/lib.rs b/piet-cairo/src/lib.rs
index b97ec16..e2985d7 100644
--- a/piet-cairo/src/lib.rs
+++ b/piet-cairo/src/lib.rs
@@ -242,12 +242,23 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
             for y in 0..height {
                 let src_off = y * bytes_per_row;
                 let dst_off = y * stride;
+                #[all(not(cfg(target_endian = "big"), cfg(target_endian = "little")))]
+                error("unsupported endian");
                 match format {
                     ImageFormat::Rgb => {
                         for x in 0..width {
-                            data[dst_off + x * 4 + 0] = buf[src_off + x * 3 + 2];
-                            data[dst_off + x * 4 + 1] = buf[src_off + x * 3 + 1];
-                            data[dst_off + x * 4 + 2] = buf[src_off + x * 3 + 0];
+                            #[cfg(target_endian = "big")]
+                            {
+                                data[dst_off + x * 4 + 3] = buf[src_off + x * 3 + 2];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x * 3 + 1];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x * 3 + 0];
+                            }
+                            #[cfg(target_endian = "little")]
+                            {
+                                data[dst_off + x * 4 + 0] = buf[src_off + x * 3 + 2];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x * 3 + 1];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x * 3 + 0];
+                            }
                         }
                     }
                     ImageFormat::RgbaPremul => {
@@ -255,10 +266,20 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
                         // hope that LLVM generates pretty good code for this.
                         // TODO: consider adding BgraPremul format.
                         for x in 0..width {
-                            data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 2];
-                            data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 1];
-                            data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
-                            data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 3];
+                            #[cfg(target_endian = "big")]
+                            {
+                                data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 2];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 1];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 0];
+                                data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 3];
+                            }
+                            #[cfg(target_endian = "little")]
+                            {
+                                data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 2];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 1];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
+                                data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 3];
+                            }
                         }
                     }
                     ImageFormat::RgbaSeparate => {
@@ -268,17 +289,36 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
                         }
                         for x in 0..width {
                             let a = buf[src_off + x * 4 + 3];
-                            data[dst_off + x * 4 + 0] = premul(buf[src_off + x * 4 + 2], a);
-                            data[dst_off + x * 4 + 1] = premul(buf[src_off + x * 4 + 1], a);
-                            data[dst_off + x * 4 + 2] = premul(buf[src_off + x * 4 + 0], a);
-                            data[dst_off + x * 4 + 3] = a;
+                            #[cfg(target_endian = "big")]
+                            {
+                                data[dst_off + x * 4 + 3] = premul(buf[src_off + x * 4 + 2], a);
+                                data[dst_off + x * 4 + 2] = premul(buf[src_off + x * 4 + 1], a);
+                                data[dst_off + x * 4 + 1] = premul(buf[src_off + x * 4 + 0], a);
+                                data[dst_off + x * 4 + 0] = a;
+                            }
+                            #[cfg(target_endian = "little")]
+                            {
+                                data[dst_off + x * 4 + 0] = premul(buf[src_off + x * 4 + 2], a);
+                                data[dst_off + x * 4 + 1] = premul(buf[src_off + x * 4 + 1], a);
+                                data[dst_off + x * 4 + 2] = premul(buf[src_off + x * 4 + 0], a);
+                                data[dst_off + x * 4 + 3] = a;
+                            }
                         }
                     }
                     ImageFormat::Grayscale => {
                         for x in 0..width {
-                            data[dst_off + x * 4 + 0] = buf[src_off + x];
-                            data[dst_off + x * 4 + 1] = buf[src_off + x];
-                            data[dst_off + x * 4 + 2] = buf[src_off + x];
+                            #[cfg(target_endian = "big")]
+                            {
+                                data[dst_off + x * 4 + 3] = buf[src_off + x];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x];
+                            }
+                            #[cfg(target_endian = "little")]
+                            {
+                                data[dst_off + x * 4 + 0] = buf[src_off + x];
+                                data[dst_off + x * 4 + 1] = buf[src_off + x];
+                                data[dst_off + x * 4 + 2] = buf[src_off + x];
+                            }
                         }
                     }
                     _ => return Err(Error::NotSupported),
```
>  it really comes down to what the generated code looks like, 

I haven't looked at any assembly and even if I did, I am not sure I'd draw the right conclusions. I guess one would need some benchmarks instead to compare approaches.

Do you want to go down this route? If so, I think it would make more sense to add some benchmarks in a separate PR and merge them first.

Reply:
My only real concern is with ensuring we aren't slowing down the little-endian path, so I'm trying to think through what our options are. 

we might even be able to do something like,

```rust
// somewhere
const fn native_endian_offsets() -> (usize, usize, usize, usize) {
    #[cfg(target_endian = "big")]
    {
        return (3, 2, 1, 0)
    }
    return (0, 1, 2, 3)
}

// before the match arm:
let (r, g, b, a) = native_endian_offsets();

// and then our indexing code:
data[dst_off + x * 4 + r] = buf[src_off + x * 3 + 2];
data[dst_off + x * 4 + g] = buf[src_off + x * 3 + 1];
data[dst_off + x * 4 + b] = buf[src_off + x * 3 + 0];
```

or something? Just thinking out loud.

Reply:
I tried looking into this a bit more. Here is a GodBolt-experiment:

https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlJvoCqAZ0wAFAB4gA5AAYppAFZdSrZrVDEArkIKkR7ZATx1KmWugDCqVpoC2tZWfQAZPLUwA5OwCNMxEABMMqQADqhChMa0Vjb2ymERRnSu7l62vv5BepgGSbQMBMzEBDF2DpzZuVEFRQQpnj5%2BgcE6taVxFa3F9WkZzQCUeqiaxMgcUiGa3gDUVLTTtswA1pgA%2BniLwGskeMBubKusuwgEEN6aVCDTAKQBAGzXAKwAQpoAHE8AIqTT6MyFV1ud1smgINxe7y%2BPwA7nh0AQkNNtHgAF6Yfo3ADsz2uMgAgtNCbMSNNxNM3NMZAA6Kmw%2BEILE4/FElm/f7MJ7PMkAKmm3BuAWelK%2BNwAzJ9pucqJyedNRQKhQEvtdRUyCazCX9CjLprz%2BbchZwRSqJVKdbz5QbpkbHp8VWqNZr2ea%2BQrpkrbWLTRcXZbBcLbfbcerCddMXb8WGI3jg5MZnMFss1htmFtVrYMHgqHhMOhDsdTlLAfdOZDbT8tcxi8DQeDXh9y9M6Qirsi0Riww6iTRiKTyfNqbS4QjGcHHU3iIQ1sRgN4IGPx0TK6QF4vSSvmWvJT6XrK/YqoavF2bd7q5W6bd8j%2BOT1yz/uA1fNxr%2BkHn1jo1Hg8GE9DJwRp1nZgIErasQTBUsG2%2BUlWwiNEfn8JE3h%2BUAkJ%2BbxW2Q6YqyQjtsSPdgwQgZgEJQ9CO3FaYIE0UUAhAEAqGIVBbGIwYkVo%2BjGOYiBiDYmi6IYpiWOAPiOME7jvH6V9VQIzAwRCPBxByL1sLFCwVQsd19UxTTew0jTrTuLFNOANSDLeYzJTfEM2W1XduW4GkIHEA1OH6ByvipNAQgAT1WLjbFWIQjjGed7gUpTWCpIhVncVZvB8gChAgKTrK/SN8V/f9ALnUCBRrCCISgn5JCRODMAQzCUKqyVMLwrtCT/KdVhnbxgOXddpjIaZgHQn4ZGktUoykQZWGkR55AcWR5FQaQACVtDBIRhlGTABVFTh5AIaQ5Ck0glhAR5gjGqRuEmnaZukeQhBAYJtumwY4FgJA0FsBT2DICgIFe96mloAtWB8vg8FYADiBus4LtIbx9mIHzpE20hXtsMwCAAeVoQGoawRZ1HYbG8GITBDDwAA3TAbum0hMCU5BQXGRG3AAk65BUPBvGIIofKsLAoYISdbARka%2BDoRgWHxngRcEEQJChpQKlUdQQC0HQ2e8G7IEGVAQjySmAForB1vX2HJ1gTQCa6cmJvJTHMdpymp8wekafwKgSSI6Ht%2BJwg92hnfSJpOitkm6BqYovaDqpQ8Kbo3AaAPXb0GOSmsMplC6Oo496QPBmWkYxi4UbxvOqnZqkFWCGQaZ/uAE5Aao3BCBJW4Np%2BKw3pBvx1rc6YFp0LaLr2g6jpUaQztIKbWbL67btIe7dqLqQLYnqHp7nwfBnJ8GohAbggA%3D

This is just the `ImageFormat::Rgb`-case. `make_image_original_light` is supposed to be the original version and `make_image_modified_light` is supposed to represent the version from this PR.

Both versions are bad.

The original version:
- There is a out-of-bounds-check for every single indexing into the slice, i.e. six bounds check per loop iteration.
- The data is really copied around byte by byte.
- For example, just the line `data[x * 4 + 2] = buf[x * 3 + 0];` results in:
```
        # First bounds check
        cmp     r10, rsi
        jae     .LBB0_16
        # Second bounds check
        cmp     r11, rcx
        jae     .LBB0_17
        # Copy [rdi + r9 - 2] to [rdx + r11]
        movzx   eax, byte ptr [rdi + r9 - 2]
        mov     byte ptr [rdx + r11], al
```
So, we have two length checks. Afterwards, a single byte is copied. It looks like the indexing was rewritten by the compiler and thus the offsets are different. At the end of each loop iteration, we have the following, which seems mostly consistent with the expected loop variables:
```
        add     r11, 4
        add     r9, 3
        add     r8, -1
        jne     .LBB0_2
```

"My" version is slightly better.
- There are only three bounds checks per pixel.
- The data is copied per-byte into registers, then put into `ebp` and written to the destination in a single memory access.
- The whole loop is:
```
.LBB1_2:
        # I guess these are three bounds checks for buf[x*3+2], data[x*4] and data[(x+1)*4], but have not checked throughly
        cmp     rax, rsi
        jae     .LBB1_7
        cmp     r10, r11
        je      .LBB1_8
        cmp     r9, rcx
        ja      .LBB1_9
       # Get the three bytes into ebx, ebp, r14d
        movzx   ebx, byte ptr [rdi + rax]
        movzx   ebp, byte ptr [rdi + rax - 1]
        movzx   r14d, byte ptr [rdi + rax - 2]
        # Shift things around
        shl     ebx, 16
        shl     ebp, 8
        # Or everything together
        or      ebp, ebx
        or      ebp, r14d
        # Write the resulting u32
        mov     dword ptr [rdx + 4*r10], ebp
       # Increment counters for the next loop iteration
        add     r10, 1
        add     r9, 4
        add     rax, 3
        # Jump to next loop iteration
        cmp     r8, r10
        jne     .LBB1_2
```

> My only real concern is with ensuring we aren't slowing down the little-endian path,

I think this is right now mostly a theoretical concern. "The code is different, so it might be slower." Should I start working on a benchmark? As I wrote above: 

>> it really comes down to what the generated code looks like,
>
> I haven't looked at any assembly and even if I did, I am not sure I'd draw the right conclusions. I guess one would need some benchmarks instead to compare approaches.
>
> Do you want to go down this route? If so, I think it would make more sense to add some benchmarks in a separate PR and merge them first.

Reply:
Or should I just go with your suggested `native_endian_offsets()` function and that's good enough?